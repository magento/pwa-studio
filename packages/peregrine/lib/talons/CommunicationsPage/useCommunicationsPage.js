import { useCallback, useMemo } from 'react';
import { useMutation, useQuery } from '@apollo/client';

import mergeOperations from '../../util/shallowMerge';
import { useUserContext } from '../../context/user';

import DEFAULT_OPERATIONS from './communicationsPage.gql';

export const useCommunicationsPage = props => {
    const { afterSubmit } = props;

    const operations = mergeOperations(DEFAULT_OPERATIONS, props.operations);
    const {
        getCustomerSubscriptionQuery,
        setNewsletterSubscriptionMutation
    } = operations;

    const [{ isSignedIn }] = useUserContext();

    const { data: subscriptionData, error: subscriptionDataError } = useQuery(
        getCustomerSubscriptionQuery,
        { skip: !isSignedIn }
    );

    const initialValues = useMemo(() => {
        if (subscriptionData) {
            return { isSubscribed: subscriptionData.customer.is_subscribed };
        }
    }, [subscriptionData]);

    const [
        setNewsletterSubscription,
        { error: setNewsletterSubscriptionError, loading: isSubmitting }
    ] = useMutation(setNewsletterSubscriptionMutation);

    const handleSubmit = useCallback(
        async formValues => {
            try {
                await setNewsletterSubscription({
                    variables: formValues
                });
            } catch {
                // we have an onError link that logs errors, and FormError already renders this error, so just return
                // to avoid triggering the success callback
                return;
            }
            if (afterSubmit) {
                afterSubmit();
            }
        },
        [setNewsletterSubscription, afterSubmit]
    );

    return {
        formErrors: [setNewsletterSubscriptionError, subscriptionDataError],
        initialValues,
        handleSubmit,
        isDisabled: isSubmitting
    };
};
