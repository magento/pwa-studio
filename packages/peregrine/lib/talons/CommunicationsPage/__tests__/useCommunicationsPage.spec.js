import React from 'react';
import { useQuery } from '@apollo/client';

import createTestInstance from '../../../util/createTestInstance';
import { useCommunicationsPage } from '../useCommunicationsPage';

const mockSetNewsletterSubscription = jest.fn();

jest.mock('@magento/peregrine/lib/context/user', () => {
    const state = {
        isSignedIn: true
    };
    const api = {};
    const useUserContext = jest.fn(() => [state, api]);

    return { useUserContext };
});

jest.mock('@apollo/client', () => {
    const apolloClient = jest.requireActual('@apollo/client');

    return {
        ...apolloClient,
        useMutation: jest.fn().mockImplementation(mutation => {
            if (mutation === 'setNewsletterSubscriptionMutation')
                return [mockSetNewsletterSubscription, { loading: false }];

            return;
        }),
        useQuery: jest.fn().mockReturnValue({
            data: {
                customer: {
                    id: null,
                    is_subscribed: false
                }
            },
            error: null,
            loading: false
        })
    };
});

const Component = props => {
    const talonProps = useCommunicationsPage(props);
    return <i talonProps={talonProps} />;
};

const afterSubmit = jest.fn();

const mockProps = {
    afterSubmit,
    operations: {
        getCustomerSubscriptionQuery: 'getCustomerSubscriptionQuery',
        setNewsletterSubscriptionMutation: 'setNewsletterSubscriptionMutation'
    }
};

test('return correct shape while data is loading', () => {
    useQuery.mockReturnValueOnce({
        loading: true
    });

    const tree = createTestInstance(<Component {...mockProps} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    expect(talonProps).toMatchSnapshot();
});

test('return correct shape for initial customer data', () => {
    const tree = createTestInstance(<Component {...mockProps} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    expect(talonProps).toMatchSnapshot();
});

test('return correct shape for new value and fire create mutation', async () => {
    useQuery.mockReturnValueOnce({
        data: {
            customer: {
                id: 5,
                is_subscribed: false
            }
        },
        error: null,
        loading: true
    });

    const tree = createTestInstance(<Component {...mockProps} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    expect(talonProps).toMatchSnapshot();

    const { handleSubmit } = talonProps;

    await handleSubmit({
        isSubscribed: true
    });

    expect(mockSetNewsletterSubscription).toHaveBeenCalled();
    expect(mockSetNewsletterSubscription.mock.calls[0][0]).toMatchSnapshot();
    expect(afterSubmit).toHaveBeenCalled();
});

test('does not call afterSubmit if mutation fails', async () => {
    mockSetNewsletterSubscription.mockRejectedValue('Apollo Error');

    const tree = createTestInstance(<Component {...mockProps} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;
    const { handleSubmit } = talonProps;

    await handleSubmit({
        isSubscribed: true
    });

    expect(mockSetNewsletterSubscription).toHaveBeenCalled();
    expect(afterSubmit).not.toHaveBeenCalled();
});
