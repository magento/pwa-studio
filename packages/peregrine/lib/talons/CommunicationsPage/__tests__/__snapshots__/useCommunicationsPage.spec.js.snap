// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`return correct shape for initial customer data 1`] = `
Object {
  "formErrors": Array [
    undefined,
    null,
  ],
  "handleSubmit": [Function],
  "initialValues": Object {
    "isSubscribed": false,
  },
  "isDisabled": false,
  "isSignedIn": true,
}
`;

exports[`return correct shape for new value and fire create mutation 1`] = `
Object {
  "formErrors": Array [
    undefined,
    null,
  ],
  "handleSubmit": [Function],
  "initialValues": Object {
    "isSubscribed": false,
  },
  "isDisabled": false,
  "isSignedIn": true,
}
`;

exports[`return correct shape for new value and fire create mutation 2`] = `
Object {
  "variables": Object {
    "isSubscribed": true,
  },
}
`;

exports[`return correct shape while data is loading 1`] = `
Object {
  "formErrors": Array [
    undefined,
    undefined,
  ],
  "handleSubmit": [Function],
  "initialValues": undefined,
  "isDisabled": false,
  "isSignedIn": true,
}
`;
